File: arch/core/src/main/java/com/webank/ai/fate/core/mlmodel/model/MachineLearningModel.java
/*
 * Copyright 2019 The FATE Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.webank.ai.fate.core.mlmodel.model;


import java.util.Map;
import com.google.protobuf.ByteString;
import com.webank.ai.fate.api.networking.proxy.Proxy;
import com.webank.ai.fate.core.network.grpc.client.ClientPool;
import com.webank.ai.fate.core.result.ReturnResult;
import com.webank.ai.fate.core.utils.Configuration;
import com.webank.ai.fate.core.utils.ObjectTransform;
import io.grpc.ManagedChannel;
import com.webank.ai.fate.api.networking.proxy.Proxy.Packet;
import com.webank.ai.fate.api.networking.proxy.DataTransferServiceGrpc;

public abstract class MachineLearningModel<B, X, P> implements MLModel<B, X, P>{
    private Map<String, String> modelInfo;

    @Override
    public void setModelInfo(Map<String, String> modelInfo) {
        this.modelInfo = modelInfo;
    }

    @Override
    public Map<String, String> getModelInfo() {
        return this.modelInfo;
    }

    @Override
    public abstract int initModel(B modelBuffer);

    @Override
    public abstract Map<String, Object> predict(X inputData, P predictParams);

    protected Map<String, Object> getFederatedPredict(Map<String, Object> requestData){
        Packet.Builder packetBuilder = Packet.newBuilder();
        requestData.putAll(this.modelInfo);
        requestData.put("myPartyId", Configuration.getProperty("partyId"));
        packetBuilder.setBody(Proxy.Data.newBuilder()
                        .setValue(ByteString.copyFrom(ObjectTransform.bean2Json(requestData).getBytes()))
                        .build());

        Proxy.Metadata.Builder metaDataBuilder = Proxy.Metadata.newBuilder();
        Proxy.Topic.Builder topicBuilder = Proxy.Topic.newBuilder();

        metaDataBuilder.setSrc(
                topicBuilder.setPartyId(Configuration.getProperty("partyId")).
                        setRole(this.modelInfo.get("myRole"))
                        .setName("partyName")
                        .build());
        metaDataBuilder.setDst(
                topicBuilder.setPartyId(this.modelInfo.get("partnerPartyId"))
                        .setRole("host")
                        .setName("partnerPartyName")
                        .build());
        metaDataBuilder.setCommand(Proxy.Command.newBuilder().setName("federatedPredict").build());
        packetBuilder.setHeader(metaDataBuilder.build());

        ManagedChannel channel1 = ClientPool.getChannel(Configuration.getProperty("proxy"));
        DataTransferServiceGrpc.DataTransferServiceBlockingStub stub1 = DataTransferServiceGrpc.newBlockingStub(channel1);
        Packet packet = stub1.unaryCall(packetBuilder.build());

        ReturnResult result = (ReturnResult) ObjectTransform.json2Bean(packet.getBody().getValue().toStringUtf8(), ReturnResult.class);
        return result.getData();
    }
}

File: arch/task_manager/apps/machine_learning_model.py
#
#  Copyright 2019 The FATE Authors. All Rights Reserved.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#
from arch.api.utils import file_utils
from flask import Flask, request
from arch.task_manager.settings import server_conf
from arch.task_manager.utils import publish_model
from arch.task_manager.job_manager import generate_job_id
from arch.task_manager.utils.api_utils import get_json_result, federated_api
from arch.api.version_control.control import version_history
from arch.api import eggroll
from arch.task_manager.settings import WORK_MODE, logger, SERVINGS, PARTY_ID
import json
manager = Flask(__name__)


@manager.errorhandler(500)
def internal_server_error(e):
    logger.exception(e)
    return get_json_result(status=100, msg=str(e))


@manager.route('/load', methods=['POST'])
def load_model():
    request_config = request.json
    _job_id = generate_job_id()
    all_party = set()
    for _party_ids in request_config.get('role').values():
        all_party.update(set(_party_ids))
    for _party_id in all_party:
        st, msg = federated_api(job_id=_job_id,
                                method='POST',
                                url='/model/load/do',
                                party_id=_party_id,
                                json_body=request_config)
    return get_json_result(job_id=_job_id)


@manager.route('/load/do', methods=['POST'])
def do_load_model():
    request_data = request.json
    request_data["servings"] = server_conf.get("servers", {}).get("servings", [])
    publish_model.load_model(config_data=request_data)
    return get_json_result()


@manager.route('/online', methods=['POST'])
def publish_model_online():
    request_config = request.json
    if not request_config.get('servings'):
        # get my party all servings
        request_config['servings'] = SERVINGS
    publish_model.publish_online(config_data=request_config)
    return get_json_result()


@manager.route('/version', methods=['POST'])
def query_model_version_history():
    history = version_history(data_table_namespace=request.json.get("namespace"))
    return get_json_result(data=history)
